controllers/greptimedbcluster/deployers/datanode.go

```diff
@@ -246,108 +239,7 @@ func (d *DatanodeDeployer) generateSts(cluster *v1alpha1.GreptimeDBCluster) (*ap
-				},
-				Spec: corev1.PodSpec{
-					ImagePullSecrets: cluster.Spec.Datanode.Template.ImagePullSecrets,
-					Volumes: []corev1.Volume{
-						{
-							Name: "config",
-							VolumeSource: corev1.VolumeSource{
-								EmptyDir: &corev1.EmptyDirVolumeSource{},
-							},
-						},
-					},
-					InitContainers: []corev1.Container{
-						{
-							Name:  "greptimedb-initializer",
-							Image: cluster.Spec.Initializer.Image,
-							Command: []string{
-								"greptimedb-initializer",
-							},
-							Args: []string{
-								"--config-path", defaultConfigDir + defaultConfigName,
-								"--input-config-file", defaultInputConfigFile + defaultConfigName,
-								"--datanode-rpc-port", fmt.Sprintf("%d", cluster.Spec.GRPCServicePort),
-							},
-							VolumeMounts: []corev1.VolumeMount{
-								{
-									Name:      "config",
-									MountPath: defaultConfigDir,
-								},
-							},
-							// TODO(zyy17): the datanode don't support to accept hostname.
-							Env: []corev1.EnvVar{
-								{
-									Name: "POD_IP",
-									ValueFrom: &corev1.EnvVarSource{
-										FieldRef: &corev1.ObjectFieldSelector{
-											FieldPath: "status.podIP",
-										},
-									},
-								},
-							},
-						},
-					},
-					Containers: []corev1.Container{
-						{
-							Name:      string(v1alpha1.DatanodeComponentKind),
-							Image:     cluster.Spec.Datanode.Template.MainContainer.Image,
-							Resources: *cluster.Spec.Datanode.Template.MainContainer.Resources,
-							Command:   cluster.Spec.Datanode.Template.MainContainer.Command,
-							Args:      args,
-							VolumeMounts: []corev1.VolumeMount{
-								{
-									Name:      cluster.Spec.Datanode.Storage.Name,
-									MountPath: cluster.Spec.Datanode.Storage.MountPath,
-								},
-								{
-									Name:      "config",
-									MountPath: defaultConfigDir,
-								},
-							},
-							// TODO(zyy17): the datanode don't support to accept hostname.
-							Env: []corev1.EnvVar{
-								{
-									Name: "POD_IP",
-									ValueFrom: &corev1.EnvVarSource{
-										FieldRef: &corev1.ObjectFieldSelector{
-											FieldPath: "status.podIP",
-										},
-									},
-								},
-							},
-							Ports: []corev1.ContainerPort{
-								{
-									Name:          "grpc",
-									Protocol:      corev1.ProtocolTCP,
-									ContainerPort: cluster.Spec.GRPCServicePort,
-								},
-								{
-									Name:          "http",
-									Protocol:      corev1.ProtocolTCP,
-									ContainerPort: cluster.Spec.HTTPServicePort,
-								},
-								{
-									Name:          "mysql",
-									Protocol:      corev1.ProtocolTCP,
-									ContainerPort: cluster.Spec.MySQLServicePort,
-								},
-								{
-									Name:          "postgres",
-									Protocol:      corev1.ProtocolTCP,
-									ContainerPort: cluster.Spec.PostgresServicePort,
-								},
-							},
-						},
-					},
-				},
-			},
+			Template: *d.generatePodTemplateSpec(cluster),
```

Comment:
We can use utils.GeneratePodTemplateSpec() here, not define three generatePodTemplateSpec() in datanode.go, frontend.go, meta.go?

Reply:
There is different configurations of podtemplatespec in different components(frontend/meta/datanode). I think it's hard to unify the logic.